### PR TITLE
Adding in ssh utilities for Machinist

### PR DESCRIPTION
**Pull request by aaahrens**
Adds smaller ssh-agent functionality, and the ability to write the known hosts to add a cert authority

### CODE REVIEW

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:53]: I'm a big fan of the [new error handling introduced in go1.13](https://blog.golang.org/go1.13-errors). 

Let's say the disk is full, and this error is printed in the logs. You go debug it, all you'll see is "no space on device".

My recommendation is to return:
`fmt.Errorf("could not add key %s to known_hosts file: %w", err)`

(note the '%w', and err passed as is) 
This would generate a very readable and much more informative error message (now I know the implication of no space on disk), while still allowing easy programmatic access by using the `Unwrap` or `errors.Is`, `errors.As` API.

Comment (aaahrens) [reply lib/kcerts/ssh.go]: At some point I'm going to have to bump the go version.
Been using the go's errors.Wrap instead of the fmt for a while, super nice. 
https://pkg.go.dev/github.com/pkg/errors#Wrap

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go:49]: I thought this too, but each call to WriteString actually appends a newline, I think it has to do with the epipecheck at the end of the write function. 

If you want to run the ssh_test a couple of times and triple check, go ahead. (right now it writes to the users home known_hosts file)

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go]: I actually disagree because most linters actually forbid noop continue and breaks in for loops. But implmented.

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go:21]: The reason why I'm fine with how it is now is because since each go test is essentially an isolate we can just change the homedir in each test to some tmp file and call it good.  That way it mimics somewhat its real case. 

If you think eventually kcerts is going to move to more of an ssh/ecdsa/rsa utility package I would definitely refactor into a service later on.

Comment (aaahrens) [reply lib/kcerts/ssh.go:21]: Especially if we end up isolating ALL ssh dependencies into $HOME/.enkit (which is a cool idea). That way the end user never has to deal with local .ssh screwing up with other projects.

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go:37]: See below

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go]: See below

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go]: Regarding both the os.Stat and ReadFile 

I disagree with this because there isn't a valuable reason to ignore the errors. If we ignore the errors and don't os.Stat: The linter complains, it provides less info for the end user, and can possibly obfuscate errors. Generally I like to provide as much stdout info as possible especially when mucking about files users routinely touch. 

That being said, it should try and make them if it doesn't exist, so I implemented that with proper errors.

Comment (ccontavalli) [reply lib/kcerts/ssh.go]: > I disagree with this because there isn't a valuable reason to ignore the errors.
> If we ignore the errors and don't os.Stat: The linter complains, it provides less info for the end user, and can possibly obfuscate errors. Generally I like to provide as much stdout info as possible especially when mucking about files users routinely touch.

I don't disagree with the goals, but I think the code here now is significantly more complex than it could be, and is actually making the errors harder.

Doing Stat() and then the operation (eg, Read...(), Open...(), Write...()) is common in some languages, but should generally be frowned upon as an anti-pattern: Stat() can fail (or succeed) for reasons different from that of the operation, and introduces race conditions (in between Stat() and Open(), for example, the file may have been removed). Instead of one possible error, you now have two, and both need to be managed (as, for example, a Stat() succeeding and indicating a file exists and is readable does not imply the next Open() will also succeed and allow you to read the file - fs acls, syscall filtering, ...).

Instead of using Stat() to check if a file can be written or read, just try to write or read into the file, and *check the returned error code*.

In the suggestion I had made above, I just wanted your code to work correctly in the absence of the known hosts file (perfectly legal), absence of directory, empty file, ...

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:31]: see larger explanation below - but this block, line 31 to 37, is unnecessary. If the directory does not exist, your attempt to read the `known_hosts` file will probably return "No such file or directory", which is perfectly clear.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:39]: Same for this block, line 39 to 45. I would eliminate it entirely. If the file does not exist, the read on line 47 will fail, with file does not exist.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:47]: Now: if the read fails here, it can be for many different reasons. However, the file is optional: it's ok if it doesn't exist, ssh may have never created it (fresh installation). I would change the condition to only fail if the error is != than does not exist:

```
if err != nil && !os.IsNotExist(err) {
  return fmt.Errorf("...")
}
```
and...

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:51]: here is the critical part: you must be able to write the file, and create the directory if it does not exist. I would implement it like this (note os.O_CREAT, and 0600):

```
knownHostsFile, err := os.OpenFile(knownHosts, os.O_APPEND|os.O_WRONLY|os.O_CREAT, 0600)
if err != nil {
  if !os.IsNotExist(err) {
    return fmt.Errorf("could not create known_hosts file: %w", err)
  }
  if err := os.Mkdir(knownHosts, 0700); err != nil {
    return fmt.Errorf("could not create .ssh directory: %w", err)
  }
}
```

(open returns ENOENT, equivalent to NotExist, if some of the path components do not exist, https://play.golang.org/p/sesFoHhvFyz, and `man 2 open`)

File: lib/kcerts/ssh.go
Comment (ccontavalli) [reply lib/kcerts/ssh.go:57]: It's not the WriteString, it is the caPublic that contains a `\n` at the end. `MarshalAuthorizedKey` appends a \n at the end of the string, see https://play.golang.org/p/jHPir_Je6n_S

If we're relying on that, I'd add a comment here.

// caPublic terminates with a '\n', added by ...

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go:47]: Implemented, but earlier in the code we do guarantee that it exists and has been made

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go:42]: I've implemented this in a slightly different fashion that requires less nesting.

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go:76]: The parsing is very frgaile. I had tried before to eval the output but for some reason bash kept failing on me. Didn't think of the regex, I'll try again.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:53]: this should be moved on line 49, otherwise we will leak the file descriptor if ReadAll afterward fails.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:85]: am ok fixing later, but add a TODO in case: the environment variables may be totally bogus.

Eg, `SSH_AGENT_PID` and `SSH_AUTH_SOCK` may be pointing to an `ssh-agent` that is long dead or long gone.
This happens more frequently than you think, pretty much every time you ssh on a machine with agent forwarding if you use `screen` and `tmux`: your screen session will reattach you to an old shell, likely pointing to an old agent / pid that no longer exists.

My take here: a) rather than return error, just ignore the environment `ssh-agent` if anything is fishy about it (eg, pid is not an integer), b) ideally, try to use the socket before assuming that the agent is available / alive.

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go]: nice catch, done

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go:85]: Added in todo, leaving this chain open in case you find anything else

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:93]: here and below: I would refactor the error handling so that the cache is entirely optional.

Eg, if we fail to write in cache because we are out of disk space, or disk is faulty, or ... or if we fail to read a cache file (corrupted, ...) we should still be able to ssh into other boxes.

Further, we have the same problem as above: the cache may be bogus. For example: started an agent, enjoyed for a while, machine was rebooted... now cache points to an agent that no longer exists.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:108]: rather than `os.OpenFile`, `ioutil.ReadAll` followed by `json.Unmarshal`, I recommend you use: `marshal.UnmarshalFile(filepath.Join(sshEnkitCache, "ssh.json"), &sshCache)`, from `import "github.com/enfabrica/enkit/lib/config/marshal"`.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:152]: same comment as above, recommend using `marhsal.MarshalFile(filepath.Join(sshEnkitCache, "ssh.json"), cacheToWrite)`.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:92]: you probably need a `defer store.Rollback(sshEnkitCache)` after error checking, to ensure that garbage is cleaned up if there's an error before Commit. (Rollback is guaranteed to leave the data in place)

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:80]: meta: I really like this function and the improvements you've made. I would however break this down in multiple functions. Something like this:

```
type Agent struct {
  Pid int
  Socket string
}

func FindSSHAgent(store cache.Store) (*Agent, error) {
  agent, err := LoadAgentFromEnv() // Parses env variables to find an agent.
  if err == nil && agent.IsValid() {
    return agent, nil
  }

  agent, cache, err = LoadAgentFromCache(store)
  defer store.Rollback(cache)
  if err == nil && agent.IsValid() {
    return agent, nil
  }

  agent, err = CreateAgent()
  if err != nil {
    return err
  }

  StoreAgentInCache(agent, cache)
  return nil
}
```

also: both re-using an existing agent or loading from cache (or storing in cache) are "optimizations", the code should not fail or return error if those things fail.

Comment (ccontavalli) [reply lib/kcerts/ssh.go:80]: By the way: if you do this, the ttl parameter is no longer needed? Basically: if you find a working agent, no matter what, you can reuse it.

Comment (ccontavalli) [reply lib/kcerts/ssh.go:80]: `IsValid` could be implemented as something like:
```
	conn, err := net.Dial("unix", agent.Socket)
	defer conn.Close()
        return err != nil
```

you could even have an `Open()` on the `Agent` object to later on use it.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:111]: I would remove this block, from line 111 to 115. It's none of the caller's business to cleanup after FetchSSHAgentFromCache own mess, and ... due to how the cache is structured, sure enough this file will be overwritten the first time we succeed writing in the cache, don't bother removing the entry.

(fun fact: if one was using nfs or bind mounts, the cached file may be valid in one terminal, invalid in another - I recommend being lazy here)

File: lib/kcerts/cache.go
Comment (ccontavalli) [inline lib/kcerts/cache.go:27]: Following my suggestion below about a "Close" method:

`&SSHAgent{Close: func() {}}`

(when the agent is read from cache, there's nothing to close)

File: lib/kcerts/cache.go
Comment (ccontavalli) [inline lib/kcerts/cache.go:40]: following my advice below for a `Close` member... remove this line.
(if we failed to write the agent file, it has to be closed as normal)

File: lib/kcerts/cache.go
Comment (ccontavalli) [inline lib/kcerts/cache.go:46]: following my advice below for a Close member... remove this line.
(if we failed to write the agent file, it has to be closed as normal)

File: lib/kcerts/cache.go
Comment (ccontavalli) [inline lib/kcerts/cache.go:51]: change this condition to be:
```
if err == nil {
  // Now that the agent is cached, it can be used in the future or in other terminals.
  // Don't close it.
  agent.Close = func() {}
}
```

File: lib/kcerts/cache.go
Comment (ccontavalli) [inline lib/kcerts/cache.go:57]: See my suggestion below: I would remove this method, as I would not invoke it from the code.

If you want to keep it (I'm ok with it, for API completeness), I strongly recommend using `Exists` instead of `Get`.
(due to how the cache was designed for atomicity, `Get` either returns an existing entry, or creates a new one in one go - it's a bit wasteful to create an entry just to remove it 3 lines below).

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:94]: If you follow my advice of creating a Close attribute above, CloseIfNotFailed can be removed entirely, get rid of it.
You'll be left with: `Close` attribute (that depending on how the agent was created it'll close it in different ways) and `Kill`, to actually kill the agent.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:137]: `&SSHAgent{Close: func () {}, PID: pid, Socket: envSSHSock}`

(nothing to do if we try to close this agent, does not belong to us)

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:163]: Change this in:
```
agent := &SSHAgent{Socket: rawSock, PID: pid}
agent.Close = func () { agent.Kill() }
return agent
```

(when this agent is closed, we need to kill the PID)

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go:93]: The sighup is usually the fastest way I've found for ssh-add to know that the agent got killed, but that's pretty anecdotal

File: lib/kcerts/ssh.go
Comment (aaahrens) [reply lib/kcerts/ssh.go:52]: good catch

File: lib/kcerts/cache.go
Comment (ccontavalli) [inline lib/kcerts/cache.go:40]: no, do the opposite here: if writing succeeds, than set the "Close" to a noop function. Otherwise, leave it as is.
In the function that creates an agent, set the "Close" to "Kill" the agent.

Rationale:
As the code is written now, if one uses the API in different ways, it can lead to buggy / wrong results.

For example, let's say I change the code (or use the functions) so that `WriteAgentToCache` is always invoked, even when I've loaded the agent from the environment, or when I've read it from cache. If `WriteAgentToCache` fails, we end up killing *innocent* agents!! Some poor agents that don't belong to us.
(from documentation, this use of the functions is perfectly legitimate).

Another example: I create an agent invoking the function below. Then I never write to cache. I call Close() assuming the agent will be terminated, but it won't! As it's WriteAgentToCache that kills the agent in case of error.

Basically:
- the function that creates the SSHAgent owns the agent, and knows how to kill it.
- WriteAgentToCache, if it succeeds, should prevent the agent from being killed, as it's now part of the cache.
if it fails, it should terminate the agent as whoever owns it intended for it to be terminated.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:102]: nit: this TODO is now addressed

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:100]: nit: comment is out of date.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:126]: nit: if you invert the if, eg `if envSSHSock == "" || envSSHPID == "" { return nil }` you reduce the newting.

File: lib/kcerts/ssh.go
Comment (ccontavalli) [inline lib/kcerts/ssh.go:48]: nit: period at end of comment.

Comment (ccontavalli) [reply lib/kcerts/ssh.go:48]: here and below.

File: lib/kcerts/cache.go
Comment (ccontavalli) [inline lib/kcerts/cache.go:33]: ping on the comment I had added here. The SSHAgentCacheInvalid here... is pretty much just a string? I would recommend removing SSHAgentCacheInvalid. 

Basically: the function returns "an agent, and an error", an error is returned if the cache is invalid (eg, line 24 is a form of invalid cache), unless it's SSHAgentNoCache.

**Review by ccontavalli — APPROVED** [review]
All nits remaining. Please address them (recommend looking at function documentation and comments in general, add period, grammar), and merge. No need to wait for further review from me.

**Pull request merged**
